database/rrddim.c

```diff
@@ -400,13 +400,37 @@ RRDDIM *rrddim_add_custom(RRDSET *st, const char *id, const char *name, collecte
     if(unlikely(rrddim_index_add(st, rd) != rd))
         error("RRDDIM: INTERNAL ERROR: attempt to index duplicate dimension '%s' on chart '%s'", rd->id, st->id);
 
+    rrdset_unlock(st);
     if(host->alarms_with_foreach || host->alarms_template_with_foreach) {
-        rrdhost_wrlock(host);
-        rrdcalc_link_to_rrddim(rd, st, host);
-
-        rrdhost_unlock(host);
+        int count = 0;
+        int hostlocked;
+        trylockagain:
+        hostlocked = pthread_rwlock_trywrlock(&host->rrdhost_rwlock);
+        if(!hostlocked)
+        {
+            rrdcalc_link_to_rrddim(rd, st, host);
```

Comment:
Here you are reusing the `rd` dimension pointer without holding the lock for chart `st`. If `rd` is deleted in the meantime it would crash. Moreover, even if you retook the chart lock, still you would be reusing the `rd` pointer which could have been freed during the period when you did not actually hold the chart lock.

We need to think this through.

Reply:
We are creating ```rd``` in this function, so we cannot delete it at this point. What we can do is to move back `rrdset_unlock` this will avoid we delete the chart. Considering that the write lock for chart is different of write lock for host, I do not think the problem was there. Right now I really believe that the error reported from Coverity was related I try to lock something that was possibly locked. I moved the function back and I did not notice any error reported.

Reply:
Well, since you moved the lock again `rd` is safe again but you still keep the wrong lock order only now you're try-locking instead of locking. So, in the previous cases of deadlock, now you're delaying and failing, which is probably why coverity does not write this down. But still, there's probably cases where we delay and fail during try-locking. Is that a problem?

Reply:
I could not detect any error while Netdata was trying to lock to create the alarm, but I had when I tried to do some changes in the previous commits. 
About your question, I thought about it and I decided to try 5 times and finally raise an error. What can happen is we do not have an alarm created, but I used Netdata simple pattern to create 11 dimensions in a chart and I did not have any error when I tried to lock, so I think this is not a problem, at least it was not a problem with the charts I tried.
After to read your message I decided to repeat a test that I did in your PR, I changed the example.python to have 160 dimensions and I created an alarm using `foreach` to create an alarm for each dimensions and again I did not have any error.

Reply:
Make sure you fix the grammar of the error message:
```
After to try creating an alarm for dimension %s on chart %s 5 times without success, I am moving in front
```
to something like:
```
Failed to create an alarm for dimension %s of chart %s 5 times. Skipping alarm.
```

Reply:
Thank you, I changed the previous message for your suggestion, I liked it so I decided to keep it .

Reply:
You should use `netdata_rwlock_trywrlock()` instead of using pthread locks directly. Also, please reduce the number of lines according to our latest code style specifications and try to use a `for()` loop instead of the `goto` if possible here.

Reply:
I executed the changes asked and Netdata continues working without raise errors and with all alarms created.

Reply:
> Also, please reduce the number of lines according to our latest code style specifications

@thiagoftsm clion can autoformat code, just use it, see `code` section

Reply:
I will check this, thanks @ilyam8!

Reply:
I'm waiting for the code style change to approve this.

Reply:
The last commit fixes the format. Thanks!